File: Lab_03/A.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringJoiner;

public class A {
    public static String[] merge(String[] part1, String[] part2) {
        String[] mergedArray = new String[part1.length + part2.length];
        int i = 0, j = 0;
        while(i < part1.length && j < part2.length) {
            if(Integer.parseInt(part1[i]) > Integer.parseInt(part2[j])) {
                Global.inversionCounter += part1.length - i;
                mergedArray[i + j] = part2[j];
                j++;
            }
            else {
                mergedArray[i + j] = part1[i];
                i++;
            }
        }

        while(i < part1.length) {
            mergedArray[i + j] = part1[i];
            i++;
        }

        while(j < part2.length) {
            mergedArray[i + j] = part2[j];
            j++;
        }
        return mergedArray;
    }

    public static String[] mergeSort(String[] arr) {
        if(arr.length == 1) {
            return arr;
        }
        int mid = arr.length / 2;
        String[] part1 = mergeSort(Arrays.copyOfRange(arr, 0, mid));
        String[] part2 = mergeSort(Arrays.copyOfRange(arr, mid, arr.length));
        return merge(part1, part2);
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringJoiner sj = new StringJoiner(" ");
        int arrLength = Integer.parseInt(br.readLine());
        String[] numStringArr = br.readLine().split(" ");
        String[] sortedArray = mergeSort(numStringArr);
        for(int i = 0; i < sortedArray.length; i++) {
            sj.add(sortedArray[i]);
        }
        System.out.println(Global.inversionCounter);
        System.out.println(sj.toString());
    }
}

class Global{
    static int inversionCounter = 0;
}

File: Lab_03/A.py
inversionCounter = 0

def merge(part1, part2):
    global inversionCounter
    mergedList = []
    i, j = 0, 0
    while i < len(part1) and j < len(part2):
        if int(part1[i]) > int(part2[j]):
            mergedList.append(part2[j])
            inversionCounter += len(part1) - i
            j += 1
        else:
            mergedList.append(part1[i])
            i += 1

    while i < len(part1):
        mergedList.append(part1[i])
        i += 1
    while j < len(part2):
        mergedList.append(part2[j])
        j += 1
    return mergedList

def mergeSort(arr):
    if len(arr) <= 1:
        return arr
    else:
        mid = len(arr)//2
        a1 = mergeSort(arr[0: mid:])
        a2 = mergeSort(arr[mid: len(arr):])
        return merge(a1, a2)
    
arrLength = int(input())
numStringArr = input().split(" ")
sortedArray = mergeSort(numStringArr)
stringRepresentation = " ".join(sortedArray)
print(inversionCounter)
print(stringRepresentation)

